File: apps/api/services/connectors/google_drive.py
"""Google Drive connector: OAuth (refresh_token), list/search files, return fragments."""

import asyncio

from google.oauth2.credentials import Credentials
from google.auth.transport.requests import Request
from googleapiclient.discovery import build
from googleapiclient.errors import HttpError

from config import settings
from services.connectors.contract import (
    ConnectorInput,
    ConnectorOutput,
    Fragment,
    SourceMetadata,
)

DRIVE_SCOPE = "https://www.googleapis.com/auth/drive.readonly"


def _credentials_from_config(config: dict):
    """Build Credentials from tenant config. Expects _credentials.refresh_token; uses app-level client id/secret."""
    creds_dict = config.get("_credentials") or config
    refresh_token = creds_dict.get("refresh_token")
    if not refresh_token:
        raise ValueError("refresh_token required")
    if not settings.google_client_id or not settings.google_client_secret:
        raise ValueError("Google OAuth not configured (GOOGLE_CLIENT_ID, GOOGLE_CLIENT_SECRET)")
    creds = Credentials(
        token=None,
        refresh_token=refresh_token,
        token_uri="https://oauth2.googleapis.com/token",
        client_id=settings.google_client_id,
        client_secret=settings.google_client_secret,
        scopes=[DRIVE_SCOPE],
    )
    creds.refresh(Request())
    return creds


def _fetch_sync(input: ConnectorInput) -> ConnectorOutput:
    limit = min(input.limits.max_fragments_per_source, 20)
    try:
        creds = _credentials_from_config(input.config)
        service = build("drive", "v3", credentials=creds)
        query = input.query_text.strip()
        if query:
            q = f"fullText contains '{query.replace(chr(39), chr(39)+chr(39))}'"
        else:
            q = None
        params = {
            "pageSize": limit,
            "fields": "files(id,name,webViewLink,mimeType)",
            "orderBy": "modifiedTime desc",
        }
        if q:
            params["q"] = q
        result = service.files().list(**params).execute()
        files = result.get("files") or []
        fragments = []
        for f in files[:limit]:
            name = f.get("name") or "(bez nazwy)"
            link = f.get("webViewLink") or f"https://drive.google.com/file/d/{f.get('id', '')}/view"
            content = f"{name}"
            fragments.append(
                Fragment(
                    content=content,
                    metadata={"id": f.get("id"), "link": link, "mimeType": f.get("mimeType")},
                )
            )
        return ConnectorOutput(
            success=True,
            fragments=fragments,
            source_metadata=SourceMetadata(
                source_id="google_drive",
                type="documents",
                title="Google Drive",
                link="https://drive.google.com",
            ),
            error=None,
        )
    except HttpError as e:
        msg = e.reason or str(e)
        if e.resp and e.resp.status in (401, 403):
            msg = "Błąd autoryzacji — połącz ponownie w panelu admin."
        return ConnectorOutput(
            success=False,
            fragments=[],
            source_metadata=SourceMetadata(
                source_id="google_drive",
                type="documents",
                title="Google Drive",
                link=None,
            ),
            error=msg[:200],
        )
    except Exception as e:
        return ConnectorOutput(
            success=False,
            fragments=[],
            source_metadata=SourceMetadata(
                source_id="google_drive",
                type="documents",
                title="Google Drive",
                link=None,
            ),
            error=str(e)[:200],
        )


class GoogleDriveAdapter:
    async def fetch(self, input: ConnectorInput) -> ConnectorOutput:
        return await asyncio.to_thread(_fetch_sync, input)

    def test_connection(self, config: dict) -> tuple[bool, str | None]:
        try:
            creds = _credentials_from_config(config)
            service = build("drive", "v3", credentials=creds)
            service.files().list(pageSize=1, fields="files(id)").execute()
            return True, None
        except Exception as e:
            msg = str(e)
            if "refresh_token" in msg or "credentials" in msg.lower():
                msg = "Nieprawidłowy refresh token — wykonaj ponownie połączenie z Google."
            return False, msg[:200]

File: apps/web/lib/server/connectors/google-drive.ts
/**
 * Google Drive connector: OAuth (refresh_token), list files, return fragments. Server-only.
 */

import { google } from "googleapis";
import type {
  ConnectorInput,
  ConnectorOutput,
  Fragment,
  SourceMetadata,
} from "./contract";

const DRIVE_SCOPE = "https://www.googleapis.com/auth/drive.readonly";

function getCredentials(config: Record<string, unknown>): { refresh_token: string } {
  const creds = (config._credentials ?? config) as Record<string, unknown>;
  const refreshToken = creds.refresh_token;
  if (!refreshToken || typeof refreshToken !== "string") {
    throw new Error("refresh_token required");
  }
  const clientId = process.env.GOOGLE_CLIENT_ID;
  const clientSecret = process.env.GOOGLE_CLIENT_SECRET;
  if (!clientId || !clientSecret) {
    throw new Error("Google OAuth not configured (GOOGLE_CLIENT_ID, GOOGLE_CLIENT_SECRET)");
  }
  return { refresh_token: refreshToken };
}

export class GoogleDriveAdapter {
  async fetch(input: ConnectorInput): Promise<ConnectorOutput> {
    const limit = Math.min(input.limits.max_fragments_per_source, 20);
    const metadata: SourceMetadata = {
      source_id: "google_drive",
      type: "documents",
      title: "Google Drive",
      link: "https://drive.google.com",
    };
    try {
      const { refresh_token } = getCredentials(input.config);
      const oauth2Client = new google.auth.OAuth2(
        process.env.GOOGLE_CLIENT_ID,
        process.env.GOOGLE_CLIENT_SECRET,
        process.env.GOOGLE_OAUTH_REDIRECT_URI
      );
      oauth2Client.setCredentials({ refresh_token });
      const drive = google.drive({ version: "v3", auth: oauth2Client });
      const rawQuery = input.query_text.trim();

      // Fragments we return are file names (`f.name`).
      // Searching `fullText` with the whole user question (e.g. "jakie mam pliki w drive")
      // often produces zero results. Prefer `name contains` with a term extracted from the question.
      const stopWords = new Set([
        // Polish common words (used in listing-type questions)
        "jakie",
        "mam",
        "moje",
        "pliki",
        "plik",
        "foldery",
        "folder",
        "w",
        "na",
        "z",
        "do",
        "dysku",
        "dysk",
        "google",
        "drive",
        "czy",
        "pokaż",
        "pokaz",
        "wyniki",
        "lista",
        "podaj",
        "wymień",
        "wymien",
      ]);

      const tokens = rawQuery
        .toLowerCase()
        .replace(/[^a-z0-9ąćęłńóśźż]+/gi, " ")
        .split(/\s+/)
        .filter(Boolean);

      const meaningful = tokens.filter((t) => {
        if (stopWords.has(t)) return false;
        // Keep digits (e.g. "2025") and longer words.
        if (/^\d+$/.test(t)) return true;
        return t.length >= 3;
      });

      const term = meaningful.slice(0, 8).join(" ");
      const escapedTerm = term.replace(/'/g, "''");
      const q = term ? `name contains '${escapedTerm}' and trashed=false` : "trashed=false";
      const res = await drive.files.list({
        pageSize: limit,
        fields: "files(id,name,webViewLink,mimeType)",
        orderBy: "modifiedTime desc",
        q,
        corpora: "allDrives",
        // Allow access to shared drives (My Drive + Shared Drives)
        supportsAllDrives: true,
        includeItemsFromAllDrives: true,
      });
      const files = res.data.files ?? [];
      const fragments: Fragment[] = files.slice(0, limit).map((f) => ({
        content: f.name ?? "(bez nazwy)",
        metadata: {
          id: f.id,
          link: f.webViewLink ?? `https://drive.google.com/file/d/${f.id}/view`,
          mimeType: f.mimeType,
        },
      }));
      return {
        success: true,
        fragments,
        source_metadata: metadata,
        error: null,
      };
    } catch (err: unknown) {
      const msg = err instanceof Error ? err.message : String(err);
      return {
        success: false,
        fragments: [],
        source_metadata: { ...metadata, link: null },
        error: msg.slice(0, 200),
      };
    }
  }

  async testConnection(config: Record<string, unknown>): Promise<{ ok: boolean; error?: string }> {
    try {
      const { refresh_token } = getCredentials(config);
      const oauth2Client = new google.auth.OAuth2(
        process.env.GOOGLE_CLIENT_ID,
        process.env.GOOGLE_CLIENT_SECRET,
        process.env.GOOGLE_OAUTH_REDIRECT_URI
      );
      oauth2Client.setCredentials({ refresh_token });
      const drive = google.drive({ version: "v3", auth: oauth2Client });
      await drive.files.list({
        pageSize: 1,
        fields: "files(id)",
        supportsAllDrives: true,
        includeItemsFromAllDrives: true,
        corpora: "allDrives",
      });
      return { ok: true };
    } catch (err: unknown) {
      const msg = err instanceof Error ? err.message : String(err);
      return { ok: false, error: msg };
    }
  }
}
